### PR TITLE
Cluster cleanup: Only create usercluster client if its actually needed

### PR DESCRIPTION
**Pull request by alvaroaleman**
We currently always require a usercluster client to even attempt to
cleanup a cluster. This means cleanup for clusters that were deleted too
quick can not work.

This PR fixes that by defering the creation of the usercluster client
until its actually being used.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

/assign @nikhita @mrIncompetent

### CODE REVIEW

**Comment by alvaroaleman** [comment]
This isn't really an isuse for 2.11 because cleanup is only needed when there is a finalizer and a finalizer is only there for things that are inside the cluster.

This is different for the credentials-secret, we always create a finalizer for it but can not clean it up now if we can't construct a usercluser client.

File: api/pkg/controller/cloud/cloud_controller.go [hunk 1]
```diff
@@ -120,8 +120,7 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluste
 		finalizers := sets.NewString(cluster.Finalizers...)
 		if finalizers.Has(kubermaticapiv1.InClusterLBCleanupFinalizer) ||
 			finalizers.Has(kubermaticapiv1.InClusterPVCleanupFinalizer) ||
-			finalizers.Has(kubermaticapiv1.NodeDeletionFinalizer) ||
-			finalizers.Has(kubermaticapiv1.CredentialsSecretsCleanupFinalizer) {
```
Comment (nikhita) [inline api/pkg/controller/cloud/cloud_controller.go:124]: Why is this removed?

File: api/pkg/clusterdeletion/deletion.go [hunk 2]
```diff
@@ -16,16 +16,16 @@ const (
 	deletedLBAnnotationName = "kubermatic.io/cleaned-up-loadbalancers"
 )
 
-func New(seedClient, userClusterClient controllerruntimeclient.Client) *Deletion {
+func New(seedClient controllerruntimeclient.Client, userClusterClientGetter func() (controllerruntimeclient.Client, error)) *Deletion {
```
Comment (mrIncompetent) [inline api/pkg/clusterdeletion/deletion.go:19]: Wouldn't it be simpler if we create 2 deletion objects?
1 for the objects inside the user cluster and 1 for the resources inside the seed

File: api/pkg/controller/cloud/cloud_controller.go (hunk 1 above)
Comment (alvaroaleman) [reply api/pkg/controller/cloud/cloud_controller.go:124]: Because the cleanup of the cloudprovider may need credentials (e.G. on AWS) but with this check  we basically don't run it before the secret is gone, deadlocking ourselves. Or am I missing something?

Also I think the deletion in the cluster controller has to somehow wait with cleaning this up until all cloudprovider resources are gone

File: api/pkg/clusterdeletion/deletion.go (hunk 2 above)
Comment (alvaroaleman) [reply api/pkg/clusterdeletion/deletion.go:19]: Which resources inside the seed exactly?

The reason we need the seed client is:

* To remove finalizers
* To store the progress of the LB cleanup in an annotation on the cluster

Comment (mrIncompetent) [reply api/pkg/clusterdeletion/deletion.go:19]: Right now we delete things inside the user cluster(LBs, PVs, etc.) & inside the seed cluster (Cloud provide secret) here.

If we would split those into 2 dedicated deletion "process objects" we would create a more readable/understandable code and the user cluster client can be initialized once (Instead of n times like now).

Comment (mrIncompetent) [reply api/pkg/clusterdeletion/deletion.go:19]: Just discussed this offline.
Either we duplicate the "i-can-delete-things-inside-the-user-cluster" check across the Kubernetes & OpenShift controllers, or we keep the "load-the-client-lazy" behaviour. Both if which i'm not a huge fan of. So lets keep with this code

**Comment by nikhita** [comment]
For cluster deletion discussion

**Comment by nikhita** [comment]
/hold

will move that to https://github.com/kubermatic/kubermatic/pull/3836

**Comment by nikhita** [comment]
dammit

/hold cancel

**Comment by floreks** [comment]
Unit test build fails with
```bash
# github.com/kubermatic/kubermatic/api/pkg/clusterdeletion [github.com/kubermatic/kubermatic/api/pkg/clusterdeletion.test]
pkg/clusterdeletion/deletion_test.go:64:19: unknown field 'userClusterClient' in struct literal of type Deletion
pkg/clusterdeletion/deletion_test.go:67:35: not enough arguments in call to d.cleanupPVCUsingPods
	have (context.Context)
	want (context.Context, "github.com/kubermatic/kubermatic/api/vendor/sigs.k8s.io/controller-runtime/pkg/client".Client)
ok  	github.com/kubermatic/kubermatic/api/cmd/image-loader	24.372s
?   	github.com/kubermatic/kubermatic/api/cmd/kubeletdnat-controller	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/kubermatic-api	[no test files]
ok  	github.com/kubermatic/kubermatic/api/cmd/kubermatic-controller-manager	1.416s
?   	github.com/kubermatic/kubermatic/api/cmd/kubermatic-operator	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/master-controller-manager	[no test files]
ok  	github.com/kubermatic/kubermatic/api/cmd/nodeport-proxy/envoy-manager	1.330s
?   	github.com/kubermatic/kubermatic/api/cmd/nodeport-proxy/lb-updater	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/owner-remover	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/projects-migrator	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/s3-exporter	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/s3-storeuploader	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/user-cluster-controller-manager	[no test files]
?   	github.com/kubermatic/kubermatic/api/cmd/userdata-openshift	[no test files]
?   	github.com/kubermatic/kubermatic/api/codegen/reconcile	[no test files]
?   	github.com/kubermatic/kubermatic/api/pkg/addon	[no test files]
ok  	github.com/kubermatic/kubermatic/api/pkg/api/v1	1.119s
?   	github.com/kubermatic/kubermatic/api/pkg/cluster/client	[no test files]
FAIL	github.com/kubermatic/kubermatic/api/pkg/clusterdeletion [build failed]
```

**Comment by alvaroaleman** [comment]
@floreks updated, can you re-lgtm?
